Add OMR init/shutdown and thread attachment

jason-hall opened the pull request:
Note that I temporarily hardcoded the path the omr static lib. I am aware this will need to be fixed, but for now, it works.

jason-hall committed:
Add OMR init/shutdown and thread attachment

charliegracie commented on js/src/jsapi.cpp:
```diff
@@ -456,6 +485,60 @@ JS::isGCEnabled()
+#if defined(OMR)
+static OMRPortLibrary portLibrary_;
+
+static omr_error_t InitializeOMR()
```
Can we call the OMR_Initialize_VM API instead?

jason-hall committed:
Stub out GC functions

jason-hall committed:
WIP startup fixes

rwy7 committed:
WIP: Add flag word to GC Cells

rwy7 committed:
Various OMR/Spidermonkey changes

jason-hall committed:
Fix ReallocateObjectBuffer

jason-hall committed:
Fix misc asserts

jason-hall committed:
Remove assrts for isTenured

jason-hall committed:
Remove asserts from shutdown and reimplement NewMemoryInfoObject()

jason-hall committed:
Change malloc to omr allocators without gc

jason-hall committed:
Change OMR_GC_AllocateNoGC to OMR_GC_Allocate and fix missing kinds

rwy7 committed:
Reenable isInsideCurrentZone temporarily

youngar committed:
Only link against libiconv on Darwin

OMR requires iconv for internationalization on many platforms.  On
Darwin these functions are provided by libiconv, while on Linux they are
provided by glibc.  We should only try to link against libiconv on
Darwin.

Signed-off-by: Andrew Young <youngar17@gmail.com>

jason-hall committed:
Fix misc typos

youngar committed:
Implement GC exclusive access

These functions are copied from MM_EnvironmentLanguageInterface to
MM_EnvironmentLanguageInterfaceImpl, with the asserts removed.

Signed-off-by: Andrew Young <youngar17@gmail.com>

youngar committed:
Update OMR submodule to the latest

Signed-off-by: Andrew Young <youngar17@gmail.com>

rwy7 committed:
Implement InterpreterFrame marking

rwy7 committed:
Everyting is wrong

rwy7 committed:
Fix a few other things

jason-hall committed:
Resurrect original tracing/marking

jason-hall committed:
Implement OMR CLI root scanning

jason-hall committed:
Bug fixes in root scanning

jason-hall committed:
More bug fixes to root scanning

jason-hall committed:
Even more bug fixes to root scanning

jason-hall committed:
Still even more bug fixes to root scanning

jason-hall committed:
Fix tracing, marking, and sweeping of weak maps/caches

jason-hall committed:
Fix JSObject slots not getting traced/scanned/marked

rwy7 committed:
Move template declarations above where they are used

- fixes a compile error on OSX

rwy7 committed:
TMP: Disable unified sources

rwy7 committed:
Revert the GC Policy code changes

rwy7 committed:
Add missing header for non-unified builds

rwy7 committed:
Include the internal gc policy header in the glue

rwy7 committed:
Move the gecko config to a header

jason-hall committed:
Fix marking of rooted jsid

jason-hall committed:
Restore cell hashing

jason-hall committed:
WIP Fix for initialShapes table

jason-hall committed:
Fix IsMarked and IsAboutToBeFinalized checks

jason-hall committed:
Fix stubbed out AutoSuppressGC

jason-hall committed:
Mark and trace TaggedProto's

jason-hall committed:
Fix LazyScript allocation template specialization

rwy7 committed:
Revert "TMP: Disable unified sources"

This reverts commit 3ff6bb06e8c077431236017bf6c0ef9387e01271.

rwy7 committed:
Only set the flags if we succeed to allocate

rwy7 committed:
Pull the zone from the GC helper

rwy7 committed:
Mark all Atoms--do not collect them

rwy7 committed:
WIP! More sweeping code

rwy7 committed:
WIP! Typo

rwy7 committed:
WIP! More zone sweeping methods

rwy7 committed:
WIP! Grab exclusive lock for sweeping the symbol table

rwy7 committed:
WIP! Kill stubbed zone member

rwy7 committed:
WIP! Add back missing PurgeJIT function

rwy7 committed:
WIP! Add missing finalize callback implementation

rwy7 committed:
WIP! Fix typo/missing sweep call

jason-hall committed:
Fix finalization callbacks

jason-hall committed:
Implemented IsAboutToBeFinalizedDuringSweep

jason-hall committed:
Fixed JS::Zone::sweepUniqueIds

jason-hall committed:
Fix init gcWeakKeys

jason-hall committed:
Remove invalid asserts

jason-hall committed:
Reimplement freeAllLifoBlocksAfterSweeping

jason-hall committed:
Fix compartments not added to list in zone

jason-hall committed:
Reimplement broken zone and compartment iteration

jason-hall committed:
Fix ObjectGroup sweeping and finalization

jason-hall committed:
Fix Shape and JSScript sweeping in finalization

jason-hall committed:
Fix unimplemented TenuredCell::isMarked

jason-hall committed:
Clear NewObjectCache on gc

jason-hall committed:
Fix missing initialization for Zone::typeDescrObjects

jason-hall committed:
Remove invalid nursery assertion

jason-hall committed:
Reimplement gc::SystemPageSize()

jason-hall committed:
Reimplement NurseryAwareHashMap as simple HashMap

jason-hall committed:
Fix broken ReallocateObjectBuffer

jason-hall committed:
Reimplement memory mapping finctions in Memory.cpp

jason-hall committed:
Fix leaked NativeObject slots

jason-hall committed:
Remove GC on every allocate
